kartothek/io_components/metapartition.py

```diff
@@ -1573,24 +1580,46 @@ def parse_input_to_metapartition(obj, metadata_version=None):
             return MetaPartition(label=None, metadata_version=metadata_version)
         first_element = obj.pop()
         mp = parse_input_to_metapartition(
-            obj=first_element, metadata_version=metadata_version
+            obj=first_element,
+            metadata_version=metadata_version,
+            expected_secondary_indices=expected_secondary_indices,
         )
         for mp_in in obj:
             mp = mp.add_metapartition(
                 parse_input_to_metapartition(
-                    obj=mp_in, metadata_version=metadata_version
+                    obj=mp_in,
+                    metadata_version=metadata_version,
+                    expected_secondary_indices=expected_secondary_indices,
                 )
             )
     elif isinstance(obj, dict):
         if isinstance(obj["data"], list):
             data = dict(obj["data"])
         else:
             data = obj["data"]
+
+        indices = obj.get("indices", {})
+        if indices:
+            warnings.warn(
+                "The explicit input of indices using the `indices` key is deprecated."
+                'Use the `secondary_indices` keyword argument of "write" and "update" functions instead.',
+                DeprecationWarning,
+            )
+        if hasattr(expected_secondary_indices, "__iter__"):
```

Comment:
Ideally, you shouldn't check for the `__iter__` attribute but be rather explicit about what you expect.
For instance a dict would be acceptable for you check as well (although I'd consider this bad input)
```python
In [1]: hasattr({}, "__iter__")
Out[1]: True
```

Reply:
I "expect" a set, since it cannot contain duplicates and will be used as a set anyways. But I thought "duck typing" would make it simple here for the users so they need not to worry about casting a list or tuple to set when making this call.

Reply:
I will change `normalize_args` so that `secondary_indices`  gets normalized to a `set`. I think that is a solution which allows for strict type parsing and flexible user input

Reply:
To remain consistent in our kwarg parsing, it would make sense that `partition_on` also supports a `set` as input (`secondary_indices` supports both `list` and `set`). I did not want to do this directly for performance reasons so I have opened #56